civis/io/_tables.py

```diff
@@ -1094,3 +1168,122 @@ def split_schema_tablename(table):
                          .format(table))
     return tuple(schema_name_tup)
+
+
+def _replace_null_column_names(column_list):
+    """Replace null names in columns from file cleaning with an appropriately
+    blank column name.
+
+    Parameters
+    ----------
+    column_list: list[dict]
+      the list of columns from file cleaning.
+
+    Returns
+    --------
+    column_list: list[dict]
+    """
+
+    new_cols = []
+    for i, col in enumerate(column_list):
+        # Avoid mutating input arguments
+        new_col = dict(col)
+        if new_col.get('name') is None:
+            new_col['name'] = 'column_{}'.format(i)
+        new_cols.append(new_col)
+    return new_cols
+
+
+def _run_cleaning(file_ids, client, need_table_columns, hidden):
+    cleaning_futures = []
+    for fid in file_ids:
+        cleaner_job = client.files.post_preprocess_csv(
+            file_id=fid,
+            in_place=False,
+            detect_table_columns=need_table_columns,
+            force_character_set_conversion=True,
+            hidden=hidden
+        )
+        cleaning_futures.append(run_job(cleaner_job.id, client=client))
+    return cleaning_futures
+
+
+def _process_cleaning_results(cleaning_futures, client, headers,
+                              need_table_columns, delimiter):
+    cleaned_file_ids = []
+
+    done, still_going = concurrent.futures.wait(
+        cleaning_futures, return_when=concurrent.futures.FIRST_COMPLETED
+    )
+
+    # Set values from first completed file cleaning - other files will be
+    # compared to this one. If inconsistencies are detected, raise an error.
+    first_completed = done.pop()
+    output_file = client.jobs.list_runs_outputs(
+            first_completed.job_id,
+            first_completed.run_id
+        )[0]
+    detected_info = client.files.get(output_file.object_id).detected_info
+    table_columns = (detected_info['tableColumns'] if need_table_columns
+                     else None)
+    if headers is None:
+        headers = detected_info['includeHeader']
+    elif headers != detected_info['includeHeader']:
+        raise CivisImportError('Mismatch between detected and provided '
+                               'headers - please ensure all imported files '
+                               'either have a header or do not.')
+    if delimiter is None:
+        delimiter = detected_info['columnDelimiter']
+    elif delimiter != detected_info['columnDelimiter']:
+        raise CivisImportError('Provided delimiter "{}" does not match '
+                               'detected delimiter for {}: "{}"'.format(
+                                   delimiter,
+                                   output_file.id,
+                                   detected_info["columnDelimiter"])
+                               )
+
+    compression = detected_info['compression']
+
+    cleaned_file_ids.append(output_file.object_id)
+
+    for result in concurrent.futures.as_completed(done | still_going):
```

Comment:
Not sure I understand what this line is doing. Is the goal to iterate over completed futures first to give the ongoing ones more time? A comment would be helpful.

Reply:
The idea is to make sure that we have accounted for all files - since `concurrent.futures.wait` returns a pair of sets, we may end up with multiple cleaners that have finished "at the same time" (i.e. when their futures poll again). By combining the set of `done`s with the set of `still_running`s, we make sure we account for all files.

I will add a comment to this effect clarifying the intent.